govcd/vm.go

```diff
@@ -850,3 +851,187 @@ func (vm *VM) SetGuestCustomizationSection(guestCustomizationSection *types.Gues
+	}
+
+	if diskData.UnitNumber < 0 {
+		return fmt.Errorf("[VM %s Id %s] disk settings unit number has to be 0 or higher", vmName, vmId)
+	}
+
+	if diskData.SizeMb < int64(0) {
+		return fmt.Errorf("[VM %s Id %s] disk settings size MB has to be 0 or higher", vmName, vmId)
+	}
+
+	if diskData.Iops != nil && *diskData.Iops < int64(0) {
+		return fmt.Errorf("[VM %s Id %s] disk settings iops has to be 0 or higher", vmName, vmId)
+	}
+
+	if diskData.ThinProvisioned == nil {
+		return fmt.Errorf("[VM %s Id %s] disk settings missing required field: thin provisioned", vmName, vmId)
+	}
+
+	if diskData.StorageProfile == nil {
+		return fmt.Errorf("[VM %s Id %s]disk settings missing required field: storage profile", vmName, vmId)
+	}
+
+	return nil
+}
+
+// GetInternalDiskById returns a valid *types.DiskSettings if it exists.
+// If it doesn't, returns nil and ErrorEntityNotFound or other err.
+func (vm *VM) GetInternalDiskById(diskId string, refresh bool) (*types.DiskSettings, error) {
+	if diskId == "" {
+		return nil, fmt.Errorf("cannot get internal disk - provided disk Id is empty")
+	}
+
+	if refresh {
+		err := vm.Refresh()
+		if err != nil {
+			return nil, fmt.Errorf("error refreshing VM: %s", err)
+		}
+	}
+
+	if vm.VM.VmSpecSection.DiskSection == nil || vm.VM.VmSpecSection.DiskSection.DiskSettings == nil ||
+		len(vm.VM.VmSpecSection.DiskSection.DiskSettings) == 0 {
+		return nil, fmt.Errorf("cannot get internal disk - VM doesn't have internal disks")
+	}
+
+	for _, diskSetting := range vm.VM.VmSpecSection.DiskSection.DiskSettings {
+		if diskSetting.DiskId == diskId {
+			return diskSetting, nil
+		}
+	}
+
+	return nil, ErrorEntityNotFound
+}
+
+// DeleteInternalDisk delete disk using provided disk ID.
+// Runs synchronously, VM is ready for another operation after this function returns.
+func (vm *VM) DeleteInternalDisk(diskId string) error {
+	err := vm.Refresh()
+	if err != nil {
+		return fmt.Errorf("error refreshing VM: %s", err)
+	}
+
+	diskSettings := vm.VM.VmSpecSection.DiskSection.DiskSettings
+	if diskSettings == nil {
+		diskSettings = []*types.DiskSettings{}
+	}
+
+	diskPlacement := -1
+	for i, diskSetting := range vm.VM.VmSpecSection.DiskSection.DiskSettings {
+		if diskSetting.DiskId == diskId {
+			diskPlacement = i
+		}
+	}
+
+	if diskPlacement == -1 {
+		return ErrorEntityNotFound
+	}
+
+	// remove disk from slice
+	diskSettings = append(diskSettings[:diskPlacement], diskSettings[diskPlacement+1:]...)
+
+	vmSpecSection := vm.VM.VmSpecSection
+	vmSpecSection.DiskSection.DiskSettings = diskSettings
+
+	_, err = vm.UpdateInternalDisks(vmSpecSection)
+	if err != nil {
+		return fmt.Errorf("error deleting VM %s internal disk %s: %s", vm.VM.Name, diskId, err)
+	}
+
+	return nil
+}
+
+// UpdateInternalDisks applies disks configuration for the VM.
+// types.VmSpecSection requires consist of all disk entities which exist and not updated,
+// as also new ones or changed ones. Returns new disk ID and error.
```

Comment:
Please rephrase. I don't understand this comment.

Reply:
updated